src/coreclr/jit/gentree.cpp

```diff
@@ -26036,6 +26036,76 @@ bool GenTreeHWIntrinsic::OperIsBitwiseHWIntrinsic() const
     return Oper == GT_AND || Oper == GT_OR || Oper == GT_XOR || Oper == GT_AND_NOT;
 }
 
+//------------------------------------------------------------------------
+// OperIsEmbRoundingEnabled: Is this HWIntrinsic a node with embedded rounding feature.
+//
+// Return Value:
+//    Whether "this" is a node with embedded rounding feature.
+//
+bool GenTreeHWIntrinsic::OperIsEmbRoundingEnabled() const
+{
+#if defined(TARGET_XARCH)
+    NamedIntrinsic intrinsicId = GetHWIntrinsicId();
+
+    if (!HWIntrinsicInfo::IsEmbRoundingCompatible(intrinsicId))
+    {
+        return false;
+    }
+
+    size_t numArgs = GetOperandCount();
+    switch (intrinsicId)
+    {
+        // these intrinsics only have the embedded rounding enabled implementation.
+        case NI_AVX512F_AddScalar:
+        case NI_AVX512F_DivideScalar:
+        case NI_AVX512F_MultiplyScalar:
+        case NI_AVX512F_SubtractScalar:
+        case NI_AVX512F_SqrtScalar:
+        {
+            return true;
+        }
+
+        case NI_AVX512F_Add:
+        case NI_AVX512F_Divide:
+        case NI_AVX512F_Multiply:
+        case NI_AVX512F_Subtract:
+
+        case NI_AVX512F_Scale:
+
+        case NI_AVX512F_ConvertScalarToVector128Double:
```

Comment:
Is this one actually correct? All of `float`, `int`, and `uint` are exactly representable and so don't round.

Reply:
Can you please elaborate more on this comment, I am not sure I fully understand it...

Reply:
The full set of API signatures for `ConvertScalarToVector128Double` are:
```csharp
public static Vector128<double> Sse2.ConvertScalarToVector128Double(Vector128<double> upper, int value); // CVTSI2SD
public static Vector128<double> Sse2.X64.ConvertScalarToVector128Double(Vector128<double> upper, long value); // CVTSI2SD
public static Vector128<double> Sse2.ConvertScalarToVector128Double(Vector128<double> upper, Vector128<float> value); // CVTSS2SD

public static Vector128<double> Avx512F.ConvertScalarToVector128Double(Vector128<double> upper, uint value); // CVTUSI2SD
public static Vector128<double> Avx512F.X64.ConvertScalarToVector128Double(Vector128<double> upper, ulong value); // CVTUSI2SD
```

Of those, the ones taking `int`, `float`, and `uint` are always exact conversions, so the instruction never takes a rounding mode.

`NI_SSE2_X64_ConvertScalarToVector128Double` and `NI_AVX512F_X64_ConvertScalarToVector128Double` are the exceptions, because they take `long` and `ulong`, respectively, which have 64-bits as compared to `double`'s mantissa which is 53-bits (52-bits explicit, 1-bit implicit)